app-store-connect/src/main.rs

```diff
@@ -0,0 +1,469 @@
+use anyhow::Result;
+use app_store_connect::bundle_api::{BundleId, BundleIdPlatform};
+use app_store_connect::certs_api::{self, Certificate, CertificateType};
+use app_store_connect::device_api::Device;
+use app_store_connect::profile_api::{Profile, ProfileType};
+use app_store_connect::{AppStoreConnectClient, UnifiedApiKey};
+use clap::{Parser, Subcommand};
+use std::io::Write;
+use std::path::PathBuf;
+
+#[derive(Parser)]
+#[clap(author, version, about, long_about = None)]
+struct Args {
+    #[clap(subcommand)]
+    command: Commands,
+}
+
+fn main() -> Result<()> {
+    env_logger::init();
+    let args = Args::parse();
+    args.command.run()
+}
+
+#[derive(Subcommand)]
+enum Commands {
+    /// Generates a PEM encoded RSA2048 signing key
+    GenerateKey {
+        /// Path to unified api key.
+        #[clap(long)]
+        api_key: PathBuf,
+        /// Certificate type can be one of development, distribution or notarization.
+        #[clap(long)]
+        r#type: String,
+        /// Path to write a new PEM encoded RSA2048 signing key
+        pem: PathBuf,
+    },
+    /// Creates a unified api key.
+    CreateApiKey {
+        /// Issuer id.
+        #[clap(long)]
+        issuer_id: String,
+        /// Key id.
+        #[clap(long)]
+        key_id: String,
+        /// Path to private key.
+        private_key: PathBuf,
+        /// Path to write a unified api key.
+        api_key: PathBuf,
+    },
+    Bundle {
+        #[clap(subcommand)]
+        command: BundleCommand,
+    },
+    Certificate {
+        #[clap(subcommand)]
+        command: CertificateCommand,
+    },
+    Device {
+        #[clap(subcommand)]
+        command: DeviceCommand,
+    },
+    Profile {
+        #[clap(subcommand)]
+        command: ProfileCommand,
+    },
+}
+
+impl Commands {
+    fn run(self) -> Result<()> {
+        match self {
+            Self::GenerateKey {
+                api_key,
+                r#type,
+                pem,
+            } => {
+                let r#type = parse_certificate_type(&r#type)?;
+                certs_api::generate_key(&api_key, r#type, &pem)?;
+            }
+            Self::CreateApiKey {
+                issuer_id,
+                key_id,
+                private_key,
+                api_key,
+            } => {
+                UnifiedApiKey::from_ecdsa_pem_path(issuer_id, key_id, private_key)?
+                    .write_json_file(api_key)?;
+            }
+            Self::Bundle { command } => command.run()?,
+            Self::Certificate { command } => command.run()?,
+            Self::Device { command } => command.run()?,
+            Self::Profile { command } => command.run()?,
+        }
+        Ok(())
+    }
+}
+
+#[derive(Subcommand)]
+enum BundleCommand {
+    Register {
+        /// Path to unified api key.
+        #[clap(long)]
+        api_key: PathBuf,
```

Comment:
Many of these (sub-)commands accept the path to an API key. I wonder if we should hoist that to a global argument in order to avoid repetition. After all, we'll probably want to eventually support multiple ways of providing the API key. Duplicating those choices on every command will be tedious and error prone.

Reply:
doesn't that enforce that the arg needs to be matched before the subcommand? having to do something like this would be weird: `app_store_connect --api-key pathtojson profile list`.

Reply:
It doesn't - at least for _global_ arguments. See https://docs.rs/clap/4.0.29/clap/builder/struct.Arg.html#method.global and the `from_global` attribute (https://docs.rs/clap/4.0.29/clap/_derive/index.html#arg-attributes).